### DISCUSSION TITLE
How to ensure that a block of code can be executed only by one request at a time in FastAPI

### DISCUSSION THREAD

**Discussion by mailtodanish** [fastapi/fastapi#8565] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
no code
```

### Description

How to ensure that a block of code can be executed only by one request at a time in FAST API

I am implementing post and calling some external tool but it takes one request at a time.

### Operating System

Linux

### FastAPI Version

0.70.0

### Python Version

Python 3.8.5

**Comment by MatthijsKok** [comment]
You can use FastAPI's [Background Tasks](https://fastapi.tiangolo.com/tutorial/background-tasks/) together with Python's [thread locks](https://docs.python.org/3/library/threading.html) to achieve this.  
See also [this stackoverflow thread](https://stackoverflow.com/questions/67193397/python-endpoint-starting-a-thread-with-locking)

**Comment by mailtodanish** [comment]
@MatthijsKok   I need response from block of code  and  will send back to request. I don't think background task will return any response.

**Comment by MatthijsKok** [comment]
Something like this?  
```
api.lock = False

@api.get("/task")
def do_task():
    if not myapp.lock:
        try:
            myapp.lock = True
            #do your external request here synchronously
            return "Task finished"
        finally:
            myapp.lock = False
    else:
        raise HTTPException(status_code=400,detail="Task already running.")
```

**Comment by mailtodanish** [comment]
thanks @MatthijsKok  api.lock will be shared in all requests? any suggest how to wait if already locked? because every request should execute block of code.
Transaction will be created in external tool but external tool accepts one request at a time.  Fast API is to create transaction in external tool using CLI.

**Comment by MatthijsKok** [comment]
What might be the best solution is to make sure that nothing is async in your api route, and then run the server with only one worker. That way you ensure that requests wait for eachother.  
`uvicorn --workers 1`

**Comment by ycd** [comment]
There is no better way to solve this problem without using a custom router.

```python
import asyncio
import time
from fastapi import FastAPI, Request, Response
from fastapi.routing import APIRoute

from typing import Callable


class CustomRoute(APIRoute):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.lock = asyncio.Lock()

    def get_route_handler(self) -> Callable:
        original_route_handler = super().get_route_handler()

        async def custom_route_handler(request: Request) -> Response:
            await self.lock.acquire()
            response: Response = await original_route_handler(request)
            self.lock.release()
            return response

        return custom_route_handler


app = FastAPI()
app.router.route_class = CustomRoute


@app.get("/hello")
async def hello():
    time.sleep(5)
    return {"hello": "world"}

```

Even if you run this with `uvicorn main:app --workers 4` this will work just fine.

**Reply by eia-imp2** [reply]
For those who want to use this solution: it doesn't release the lock if an exception was raised in the route. One can use a context manager for that:

```python
async def custom_route_handler(request: Request) -> Response:
    async with self.lock:
        response = await original_route_handler(request)
        return response
```

**Comment by mailtodanish** [comment]
@ycd please explain the approach in detail. Thanks

**Comment by ycd** [comment, marked as answer]
Read the documentation: https://fastapi.tiangolo.com/advanced/custom-request-and-route/

**Reply by Trolldemorted** [reply]
what do custom requests and routing have to do with locking a resource between multiple worker processes?

**Comment by fourdim** [comment]
Hi, streaming response often returns before all tasks are done.
Is there any way to control this behaviour so that streaming response does not release the lock before the iterator is done?

**Comment by Ruuza** [comment]
Another solution I am using is Python's `multiprocessing.Lock`:


```
import multiprocessing
from fastapi import FastAPI

app = FastAPI()
lock = multiprocessing.Lock()

@app.get("/hello")
async def hello():
    with lock:
        print("I am here")
        time.sleep(5)
        return {"hello": "world"}
```

**Reply by ckyoog** [reply]
Use asyncio.sleep() instead time.sleep()
Use asyncio lock instead of non-asyncio lock.
Otherwise, sleep in this restful api handler, will block all the other restful api calls; same, locked in this restful api handler, will block all the other restful api calls. Remember, the fastapi is a single process. All the api handlers are running in a single process. One api call can block all the other calls, unless you use async IO.

**Reply by Trolldemorted** [reply]
If you use multiple workers, not all api handlers are running in a single process, since every worker runs in its own process

**Reply by ckyoog** [reply]
That is different story. A different layer of improving the concurrency.
But I'm not talking about concurrency. I am talking about using the synchronous operation is a bad practice in the asynchronous process.
The multiple workers is not a "workaround" or "excuse" for the bad practice.
